<p style="font-size:30px; text-align:center; line-height:120%">
<br>
COMS W 4111-01, H01, V01<br>
Introduction to Databases<br>
Fall 2018 Take Home Final Exam
</p>

# Exam Overview and Instructions

- Homework assignments and exams have point values. Final grade depends on total point value. The range is 0 to 100 points. __This final exam is worth 30 points.__


- Submission:
    - The exam is due on 16-Dec-2018 at 11:59PM. __You may not use grace days.__
    - Submission:
        - Submission format is a zip file of this Jupyter Notebook with your solutions entered into the code cells or Markdown cells for each question. If you embed images or diagrams, you need to include in the zip file. 
        - You submit your zip file containing the notebook with answers on CourseWorks under "Final Examination" assignment.
        - No other formats are allowed.
        - If the notebook format is incorrect or the notebook is corrupted, the grade is 0.
        - Submissions after the due date and time are not allowed. Submissions not received on time receive a grade of 0.
    - Respect for the individual is paramount. We will accommodate special circumstances, but we must be notified and discuss __in advance.__
    

- Exam Rules:
    - Please read and review the [Academic Integrity policy and guidelines](https://www.college.columbia.edu/academics/academicintegrity), including subsections and details. This material defines the rules for this exam regardless of your school.
    - __No collaboration of any form is permitted. You may not share material of any form, including links to on-line information, information from the preparatory recitation section, suggestions or ideas, etc. You may not discuss the questions or answers with other students, TAs, friends, ...__
    - You MAY use lecture notes, material in Piazza, and any on-line information you find, but may not directly use code you find. You must cite any on-line sources in the comments Markdown cell for each questions.
    - You must privately send comments or questions to Professor Ferguson via email. If questions or comments demonstrate a need for clarification or correction, Prof. Ferguson will post an update on Piazza and send an email notification.
    - You may not discuss the exam questions or answers on Piazza, except in response to professor or CA posted comments. The CA or professor WILL post clarifying instructions on Piazza.


- Completing the exam:
    - Environment:
        - SQL:
            - You will use three MySQL databases/schema to complete the exam:
                - Lahman2017: You may use a version with any keys, indexes, etc. you have created. The only requirement is that you have not created or modified tuples that will result in your answers being different from the examples.
                - Import the _classiccars_ database/schema from https://courseworks2.columbia.edu/files/3855928/download?download_frd=1.
                - The MySQL Sakila sample database (https://dev.mysql.com/doc/sakila/en/). Please install this database.
                - You will create a database w4111final to solve problems that ask you to create new tables or schema.
            - Your code and SQL MUST use userid ```dbuser``` and password ```dbuser``` for connecting to the database.
            - You must install and use [%SQL Magic for iPtython](https://github.com/catherinedevlin/ipython-sql) for executing any SQL statements in the notebook.
            - The notebook contains several functions that help using MySQL.
        - Neo4j: You will use your instance of Neo4j for some questions. You MUST use userid=dbuser and password=dbuser for the connections. The notebook contains some helper code for Neo4j.
        - Redis: You will use Redis for some questions. The notebook contains some helper code for Redis.
        - Section 2 tests the setup of your environment. 
    - Each question starts with an explanation of the structure of the answer, e.g. prose, diagram, SQL, etc.
    - Questions requiring SQL have empty code cells where you enter your statements. Some also contain sample answers to help you determine if your answer is correct. You must show the execution of your SQL in your submission. You may use LIMIT 10 to reduce the size of results.
    - Creating test data is part of designing and implementing database solutions. You may need to create test data and add to the database for some questions. All new tables and data go into the schema you create that is your UNI.
    
    
- Points:
    - Each question in section 3 is worth two points.
    - Each question in section 4 is worth 1 point.
    - The bonus question is worth 1 point.
    - I do not want to hear any complaining about how the questions in section 3 are much more than 2x as hard as the questions in section 4. If I hear that complaining, I will make the questions in section 4 harder.

# Environment Test

## MySQL

In [3]:
import json
import pandas as pd
import matplotlib

%load_ext sql

In [4]:
%sql mysql+pymysql://dbuser:dbuser@localhost/lahman2017

/anaconda3/lib/python3.6/site-packages/pymysql/cursors.py:170: Warning: (3719, "'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous.")
  result = self._query(query)


'Connected: dbuser@lahman2017'

In [5]:
%sql select * from people where playerid='willite01'

 * mysql+pymysql://dbuser:***@localhost/lahman2017
1 rows affected.


[('willite01', '1918', '8', '30', 'USA', 'CA', 'San Diego', '2002', '7', '5', 'USA', 'FL', 'Inverness', 'Ted', 'Williams', 'Theodore Samuel', '205', '75', 'L', 'R', '1939-04-20', '1960-09-28', 'willt103', 'willite01')]

In [6]:
import pymysql.cursors
import json

pymysql_exceptions = (
    pymysql.err.IntegrityError,
    pymysql.err.MySQLError,
    pymysql.err.ProgrammingError,
    pymysql.err.InternalError,
    pymysql.err.DatabaseError,
    pymysql.err.DataError,
    pymysql.err.InterfaceError,
    pymysql.err.NotSupportedError,
    pymysql.err.OperationalError)

default_db_params = {
    "dbhost": "localhost",                    # Changeable defaults in constructor
    "port": 3306,
    "dbname": "classiccars",
    "dbuser": "dbuser",
    "dbpw": "dbuser",
    "cursorClass": pymysql.cursors.DictCursor,        # Default setting for DB connections
    "charset":  'utf8mb4'                             # Do not change
}


def get_new_connection(params=default_db_params):
    cnx = pymysql.connect(
        host=params["dbhost"],
        port=params["port"],
        user=params["dbuser"],
        password=params["dbpw"],
        db=params["dbname"],
        charset=params["charset"],
        cursorclass=params["cursorClass"])
    return cnx


def run_q(cnx, q, args, fetch=False, commit=True, cursor=None):
    """
    :param cnx: The database connection to use.
    :param q: The query string to run.
    :param args: Parameters to insert into query template if q is a template.
    :param fetch: True if this query produces a result and the function should perform and return fetchall()
    :return:
    """
    #debug_message("run_q: q = " + q)
    #ut.debug_message("Q = " + q)
    #ut.debug_message("Args = ", args)

    result = None

    try:
        if cursor is None:
            cnx = get_new_connection()
            cursor = cnx.cursor()

        result = cursor.execute(q, args)
        if fetch:
            result = cursor.fetchall()
        if commit:
            cnx.commit()
    except pymysql_exceptions as original_e:
        #print("dffutils.run_q got exception = ", original_e)
        raise(original_e)

    return result

In [7]:
q = "SELECT * from lahman2017.people where playerID=%s"
result = run_q(None, q, 'willite01', fetch=True, commit=True, cursor=None)
print("Result = ", result)

Result =  [{'playerID': 'willite01', 'birthYear': '1918', 'birthMonth': '8', 'birthDay': '30', 'birthCountry': 'USA', 'birthState': 'CA', 'birthCity': 'San Diego', 'deathYear': '2002', 'deathMonth': '7', 'deathDay': '5', 'deathCountry': 'USA', 'deathState': 'FL', 'deathCity': 'Inverness', 'nameFirst': 'Ted', 'nameLast': 'Williams', 'nameGiven': 'Theodore Samuel', 'weight': '205', 'height': '75', 'bats': 'L', 'throws': 'R', 'debut': '1939-04-20', 'finalGame': '1960-09-28', 'retroID': 'willt103', 'bbrefID': 'willite01'}]


## Neo4j

In [11]:
import py2neo

from py2neo import Graph, NodeMatcher, Node, Relationship, RelationshipMatcher

_graph = Graph(secure=False,
                    auth=("dbuser", "dbuser"),
                    bolt=True,
                    host="localhost",
                    port=7687)

_node_matcher = NodeMatcher(_graph)
_relationship_matcher = RelationshipMatcher(_graph)

r_info = {}
r_movies = []
def final_exam():
    global r_info
    global r_movies
    tom = _node_matcher.match("Person", name="Tom Hanks").first()
    print("Tom = ", tom)
    r_info['actor'] = { "name": tom['name'], "yob" : tom['born']}
    print(r_info)

    actors = set()
    actors.add(tom)
    movies = _relationship_matcher.match(nodes=actors, r_type="ACTED_IN")

    for m in movies:
        #print("Movie = ", m.end_node['title'], "; Roles = ", ", ".join(m['roles']))
        r_movies.append(m.end_node['title'])
        
    print (r_movies)
    

In [12]:
final_exam()

Tom =  None


TypeError: 'NoneType' object is not subscriptable

## Redis

In [13]:
import redis

redis_client = redis.StrictRedis(
    host='localhost',
    port=6379,
    charset="utf-8", decode_responses=True)

redis_client.hmset("actor:tom_hanks", r_info['actor'])

for m in r_movies:
    redis_client.lpush("actor:tom_hanks:movies", m)

KeyError: 'actor'

In [40]:
actor = redis_client.hgetall("actor:tom_hanks")
print("Actor = ", actor)
movies = redis_client.lrange("actor:tom_hanks:movies", 0, -1)
print(movies)

Actor =  {'name': 'Tom Hanks', 'yob': '1956'}
['Cloud Atlas', 'That Thing You Do', 'Sleepless in Seattle', 'Joe Versus the Volcano', "You've Got Mail", 'The Green Mile', 'The Da Vinci Code', 'A League of Their Own', 'Cast Away', 'Apollo 13', "Charlie Wilson's War", 'The Polar Express']


# Create, Retrieve and Update Data Questions

## Returning JSON Data (Question 3.1)

[JSON](https://www.json.org/) is an extremely common form for returning data to applications making API or [REST](https://en.wikipedia.org/wiki/Representational_state_transfer) API calls. Write a function that uses a single SQL statement to return a subset of the data comprising an order form for the _classiccars_ database.

An example of the answer for the order with orderNumber 10100 is below.

Your function's result must exactly match the field names, values, types and order as the example, except for indentation. The field names do not match the column names exactly, and you must figure out the specific columns and tables to use by examining the data in the database.

Your function MUST use EXACTLY one SQL statement. The statement MUST use exactly one JOIN and may use several sub-queries.

In [10]:
def get_order(order_no):
    q = "select\
            orders.orderNumber as order_number, orders.orderDate as order_date,\
            (select customers.customerNumber as customer_number from customers where \
                customers.customerNumber=orders.customerNumber) as customer_number,\
            (select customers.customerName as customer_name from customers where \
                customers.customerNumber=orders.customerNumber) as customer_name,\
            orderdetails.productCode as product_code,\
            (select productName as product_name from products where \
            products.productCode=orderdetails.productCode) as product_name,\
            orderdetails.quantityOrdered as quantity, orderdetails.priceEach as unit_price,\
            orderdetails.orderLineNumber as line_number \
            from orders join orderdetails on orders.orderNumber=orderdetails.orderNumber \
            where orders.orderNumber=%s \
            order by orderLineNumber;"
    
    cnx = get_new_connection()
    result = run_q(cnx, q, int(order_no), fetch = True, commit = True)
    
    if len(result) >= 0:
        final_result = {}
        final_result['customer_name'] = result[0]['customer_name']
        final_result['customer_number'] = result[0]['customer_number']
        final_result['order_number'] = result[0]['order_number']
        final_result['order_date'] = result[0]['order_date']
        
        final_result['order_items'] = []
        for r in result:
            new_item = {}
            new_item['No.'] = r['line_number']
            new_item['product_code'] = r['product_code']
            new_item['product_name'] = r['product_name']
            new_item['quantity'] = r['quantity']
            new_item['unit_price'] = float(r['unit_price'])
            
            final_result['order_items'].append(new_item)
            
    return final_result

In [11]:
def testq1():
    result = get_order(10100)
    print("Result = ", json.dumps(result, indent=2, default=str))


testq1()

Result =  {
  "customer_name": "Online Diecast Creations Co.",
  "customer_number": 363,
  "order_number": 10100,
  "order_date": "2003-01-06",
  "order_items": [
    {
      "No.": 1,
      "product_code": "S24_3969",
      "product_name": "1936 Mercedes Benz 500k Roadster",
      "quantity": 49,
      "unit_price": 35.29
    },
    {
      "No.": 2,
      "product_code": "S18_2248",
      "product_name": "1911 Ford Town Car",
      "quantity": 50,
      "unit_price": 55.09
    },
    {
      "No.": 3,
      "product_code": "S18_1749",
      "product_name": "1917 Grand Touring Sedan",
      "quantity": 30,
      "unit_price": 136.0
    },
    {
      "No.": 4,
      "product_code": "S18_4409",
      "product_name": "1932 Alfa Romeo 8C2300 Spider Sport",
      "quantity": 22,
      "unit_price": 75.46
    }
  ]
}


## Player Price Performance (Question 3.2)


HW5 focuses on determining the number of games a team will win as a function of performance summary information, e.g. on-base percentage, earned run average. We can use SQL JOINs, aggregation and functions to compute the corresponding metrics for individual players. Producing a winning team boils down to finding a set of players whose combined individual performance produces winning metrics for the team.

The situation is a little more complicated, however. There are constraints:
1. Acquiring the best performers may be prohibitively expensive.
2. A team needs players in various positions or roles, e.g. catchers, pitchers, short stops, etc.

One approach to making progress relative to the constraints is to compute a player's performance per dollar of salary. This problem is a very simple first step focusing on [catchers](https://en.wikipedia.org/wiki/Catcher).

We first have to determine which players have played as catchers. We can determine this by the _POS_ column of the _Fielding_ table. The position code for a catcher is "C." The column _InnOuts_ defines for how many "outs" a player played a position in a year. Since there are 27 outs per side per games, _sum(InnOuts)/27_ approximates the number of games a player played catcher. To identity catchers,
- We only consider _Fielding_ entries for 1950 or later.
- We only consider a player a potential catcher if the player's career _sum(InnOuts)/27_ is greater than or equal to 162.

We will focus on offensive matrics, specifically on-base percentage. Computing offensive price effectiveness requires information from three tables:
- Batting:
    - total_hits is sum(h) for the player's entire career.
    - total_abs is sum(ab) for the player's entire career.
    - batting_average is total_hits/total_abs
- Fielding:
    - total_outs is the career sum(InnOuts) for a player when catching, e.g. _Fielding.pos='C'.
    - total\_gs is the sum of gs (games started) as a catcher.
    - games is the sum of g.
    - total_outs is the sum of InnOuts
    - computed_games is sum(InnOuts)/27
- Salary:
    - salary_seasons is the number of seasons for which we have salary information. __We only consider players in the answer for whom we have at least 3 seasons of salary information.__
    - avg_salary is the average of the salary per year.
    
We derive an offensive performance/salary dollar as $ba\_dollar$ = $batting\_average/avg\_salary*10000$. We multiple to 10,000 to have larger numbers for comparison. Using the rules above, write a single SQL query that produces the information in the table below. Your initial results should be ordered by batting_average descending. Just show the first 10 rows.
    

In [15]:
q = """
select c.*,\
        (select nameLast from People where People.playerID=c.playerID) as last_name,
        (select nameFirst from People where People.playerID=c.playerID) as first_name,
        d.avg_salary, d.salary_seasons, c.batting_average/d.avg_salary*10000 as ba_dollar from 
            (select a.*, b.total_hits, b.total_abs, b.batting_average from 
                (select playerID, sum(InnOuts) total_outs, sum(GS) total_gs, 
                    sum(G) as games, round(sum(InnOuts)/27) as computed_games from Fielding 
                    where POS='C' and yearID>=1950 group by playerID having computed_games>162) as a 
                join 
                (select playerID, sum(H) as total_hits, sum(AB) as total_abs, round(sum(H)/sum(AB), 3) as batting_average  
                    from Batting where yearID>=1950 and AB>0 group by playerID) as b 
                on a.playerID=b.playerID) as c 
            join 
            (select playerID, round(avg(salary)) as avg_salary, count(*) as salary_seasons from Salaries 
                group by playerID having salary_seasons>=3 and avg_salary>0) as d 
            on c.playerID=d.playerID 
            order by batting_average desc;
"""
params = default_db_params
params['dbname'] = "lahman2017"

new_c = get_new_connection(params)
df = pd.read_sql(q, new_c)

In [16]:
df.head(10)

playerID  total_outs  total_gs   games  computed_games  total_hits  \
0  piazzmi01     40665.0    1607.0  1630.0          1506.0      2127.0   
1  mauerjo01     23649.0     885.0   920.0           876.0      1986.0   
2  poseybu01     20382.0     780.0   798.0           755.0      1163.0   
3  martivi01     21603.0     828.0   858.0           800.0      2036.0   
4  sweenmi01      4629.0     167.0   201.0           171.0      1540.0   
5  rodriiv01     61044.0    2349.0  2427.0          2261.0      2844.0   
6  harpebr01     16734.0     639.0   688.0           620.0       931.0   
7  kendaja01     52435.0    1990.0  2025.0          1942.0      2195.0   
8  lopezja01     33226.0    1251.0  1351.0          1231.0      1527.0   
9  redmomi01     16085.0     606.0   687.0           596.0       649.0   

   total_abs  batting_average  last_name first_name  avg_salary  \
0     6911.0            0.308     Piazza       Mike   8011733.0   
1     6444.0            0.308      Mauer        Joe  13232692.0   
2     3772.0            0.308      Posey     Buster   9827963.0   
3     6830.0            0.298   Martinez     Victor   8071277.0   
4     5188.0            0.297    Sweeney       Mike   4549562.0   
5     9592.0            0.296  Rodriguez       Ivan   6128697.0   
6     3151.0            0.295     Harper      Brian    922188.0   
7     7627.0            0.288    Kendall      Jason   5214398.0   
8     5319.0            0.287      Lopez       Javy   4379321.0   
9     2264.0            0.287    Redmond       Mike    690385.0   

   salary_seasons  ba_dollar  
0              15   0.000384  
1              13   0.000233  
2               6   0.000313  
3              13   0.000369  
4              16   0.000653  
5              20   0.000483  
6               8   0.003199  
7              16   0.000552  
8              14   0.000655  
9              13   0.004157

In [17]:
df2 = df[['playerID', 'last_name', 'first_name', 'avg_salary', 'batting_average']].copy()
df2['ba_per_dollar']=df2['batting_average']/df2['avg_salary']*100000
df3 = df2.sort_values(by=['ba_per_dollar'], ascending=False)
df3.head(10)

playerID   last_name first_name  avg_salary  batting_average  \
220  nietoto01       Nieto        Tom    102833.0            0.205   
142  allanan01    Allanson       Andy    124062.0            0.240   
221  bilarda01  Bilardello       Dann    106667.0            0.204   
200  deckest01      Decker      Steve    137667.0            0.221   
173  gerenbo01       Geren        Bob    151250.0            0.233   
111  salasma01       Salas       Mark    189400.0            0.247   
135  marzajo01     Marzano       John    197429.0            0.241   
170  santone01  Santovenia     Nelson    195333.0            0.233   
74   levisje01       Levis      Jesse    226750.0            0.255   
192  knorrra01       Knorr      Randy    207944.0            0.226   

     ba_per_dollar  
220       0.199352  
142       0.193452  
221       0.191249  
200       0.160532  
173       0.154050  
111       0.130412  
135       0.122069  
170       0.119283  
74        0.112459  
192       0.108683

## Shaping Data for Business Intelligence (Question 3.3)

### Context

<span style="text-decoration: underline;">_OLTP_</span>

- [Online-Transaction Processing](https://en.wikipedia.org/wiki/Online_transaction_processing) and similar applications support relatively simple read and create/update/delete, e.g.
    - Get and change my travel itinerary.
    - Get a students homework and assign grade.
    - Find product in catalog and add to shopping cart for checkout.
    
    
- Data access characteristics
    - Create, update and delete mean integrity becomes a focus, which requires constraints and avoiding update anomalies.
    - Relatively simple, recurring/common retrieves means JOINs are simple and based on keys.
    
    
- The characteristics $\Rightarrow$ normalization and decomposition are best approaches.

<br>
<span style="text-decoration: underline;">_Decision Support and Analytics_</span>

- Require complex, multi-table JOIN


- Frequently compute derived columns.


- Aggregation and summary are extremely common and recurring.


- These characteristics drive
    - Big, wide, denormalized, precomputed JOINs.
    - Precomputed summary tables.
    - Summary, list tables
    
| <img src="./dss_schema.jpeg"> |
| :---: |
| [__OLTP and Decision Support Schema__](http://www.noumenal.com/marc/schema.html) |


### Motivation

- When discussing Moneyball, we discussed figuring out wins as a function to team metrics.


- A team is a combination of players, and their performance determines the team's performance.


- A player's performance data is spread over several tables:
    - Salary
    - Batting
    - Fielding
    - Pitching
    - Appearances
    
    
- We will produce a table that summarizes player's performance metrics. We will look at the following columns:
    - Batting: G, AB, R, H, 2B, 3B, HR, RBI, BB, HBP
    - Pitching: G, W, L, IPOuts, H, ER
    - Fielding: G, PO, PA, PE, IFO, IFA, IFE, OFO, OFA, OFE, CO, CA, CE
        - There are three groups of positions: outfield (OF), infield (IF), pitcher (P), catcher (C)
            - xO means InnOuts when playing in one of the groups.
            - xA means A for the position.
            - xE means E for the position.
        - Values should be NULL if the player did not play at the position.
        
        
- Simplifications:
    - We will ignore Salary and Appearances.
    - We will also ignore seasons prior to 2010.
    - Do not include batting information for a season with less than 100 at bats. Do not including batting information for a batter with less than 500 career at bats.
    - Do not include pitching information for a season with less than 50 IPOuts. Ignore pitching information for players with less than 150 career IPOuts.
    - Do not include fielding information for a player at a position if the player played less than 50 InnOuts at the position in a season. Also, ignore if the career total is less than 150 for the position.
    
    
- Produce the summary table. 
    - This should be a copy table.
    - You may create intermediate copy tables.
    - You can create your tables using the ```create table xxx as select ... from yyy```
    - Show all of you create table statements.
    

### Sample Output

In [118]:
%sql SELECT * FROM lahman2017.player_performance_summary_final limit 10;

10 rows affected.


[('arencjp01', 432.0, 1467.0, 157.0, 306.0, 63.0, 4.0, 72.0, 224.0, 82.0, 4.0, None, None, None, None, None, None, Decimal('400'), 9950.0, 200.0, 24.0, Decimal('22'), 538.0, 13.0, 2.0, None, None, None, None, None, None, None, None, Decimal('400'), 9950.0, 200.0, 24.0),
 ('avilaal01', 788.0, 2411.0, 286.0, 583.0, 124.0, 7.0, 79.0, 311.0, 396.0, 14.0, None, None, None, None, None, None, Decimal('740'), 18416.0, 317.0, 31.0, Decimal('39'), 695.0, 11.0, 1.0, None, None, None, None, None, None, None, None, Decimal('740'), 18416.0, 317.0, 31.0),
 ('barajro01', 276.0, 875.0, 88.0, 192.0, 35.0, 0.0, 39.0, 112.0, 59.0, 12.0, None, None, None, None, None, None, Decimal('283'), 7006.0, 116.0, 11.0, None, None, None, None, None, None, None, None, None, None, None, None, Decimal('283'), 7006.0, 116.0, 11.0),
 ('barnhtu01', 317.0, 989.0, 83.0, 258.0, 56.0, 3.0, 17.0, 113.0, 103.0, 24.0, None, None, None, None, None, None, Decimal('311'), 7764.0, 219.0, 10.0, None, None, None, None, None, None, None, None, None, None, None, None, Decimal('311'), 7764.0, 219.0, 10.0),
 ('buckjo01', 465.0, 1586.0, 161.0, 366.0, 66.0, 2.0, 63.0, 224.0, 148.0, 17.0, None, None, None, None, None, None, Decimal('475'), 12236.0, 189.0, 22.0, None, None, None, None, None, None, None, None, None, None, None, None, Decimal('475'), 12236.0, 189.0, 22.0),
 ('buterdr01', 376.0, 943.0, 90.0, 193.0, 41.0, 5.0, 15.0, 85.0, 61.0, 1.0, None, None, None, None, None, None, Decimal('412'), 8548.0, 159.0, 23.0, None, None, None, None, None, None, None, None, None, None, None, None, Decimal('412'), 8548.0, 159.0, 23.0),
 ('castiwe01', 564.0, 1961.0, 204.0, 515.0, 101.0, 1.0, 77.0, 271.0, 153.0, 8.0, None, None, None, None, None, None, Decimal('558'), 14330.0, 356.0, 43.0, None, None, None, None, None, None, None, None, None, None, None, None, Decimal('558'), 14330.0, 356.0, 43.0),
 ('castrja01', 727.0, 2374.0, 289.0, 555.0, 136.0, 9.0, 72.0, 259.0, 260.0, 10.0, None, None, None, None, None, None, Decimal('680'), 16965.0, 322.0, 25.0, None, None, None, None, None, None, None, None, None, None, None, None, Decimal('680'), 16965.0, 322.0, 25.0),
 ('cervefr01', 497.0, 1578.0, 191.0, 434.0, 70.0, 12.0, 19.0, 180.0, 187.0, 3.0, None, None, None, None, None, None, Decimal('491'), 12238.0, 289.0, 43.0, None, None, None, None, None, None, None, None, None, None, None, None, Decimal('491'), 12238.0, 289.0, 43.0),
 ('chiriro01', 316.0, 949.0, 136.0, 227.0, 55.0, 2.0, 49.0, 132.0, 94.0, 1.0, None, None, None, None, None, None, Decimal('330'), 8035.0, 131.0, 17.0, None, None, None, None, None, None, None, None, None, None, None, None, Decimal('330'), 8035.0, 131.0, 17.0)]

### Comments

- If I got that correct, it was "Festivus" miracle.


- How did I try to accomplish this task?

| <img src="./wide_flat_tables.jpeg" > |
| :---: |
| __SQL Script and Operations__ |

 - If I drew that correctly, it was a miracle.

### Answer

- (JUST KIDDING. See below.) Put you SQL script and some tests in the cells below.

### Just Kidding

- The diagram below shows a subset of the SQL script you should implement.
    - Compute player total pitching performance using the rules above.
    - Compute player total batter performance using the rules above.
    - Perform full outter join to produce final result.


- Try to get approximately the correct but do not worry too much.


- The hideous scipt that I wrote is below.  You can use and simplify for simpler answer to the question.

| <img src="./wide_flat_real_assignment.jpeg" > |
| :---: |
| __SQL Script and Operations__ |

Hideous Script:

```
create table batting_summary_final as SELECT
    playerid, sum(g) as g, sum(ab) as ab, sum(r) as r, sum(h) as h, sum(`2b`) as b2, sum(`3b`) as b3,
    sum(hr) as hr, sum(rbi) as rbi, sum(bb) as bb, sum(ibb) as ibb
    FROM lahman2017.batting where yearid >= 2010 and ab >= 100
    group by playerid
    having ab >= 500;

create table pitching_summary_final as SELECT
    playerid, sum(g) as g, sum(w) as w, sum(l) as l, sum(IPOuts) as ipouts, sum(h) as h, sum(er) as er
FROM lahman2017.Pitching
where yearid >= 2010 and ipouts >= 50
group by playerid
having sum(ipouts) >= 150;

create table fielding_summary_final as SELECT
    playerid, pos, sum(g) as g, sum(innouts) as innouts, sum(a) as a, sum(e) as e
FROM lahman2017.fielding
    where innouts >= 50 and yearid >= 2010
group by playerid, pos
having sum(innouts) >= 150;

drop table if exists t_if_fielding_summary_fine;

create temporary table t_if_fielding_summary_final as SELECT
	playerid, sum(g) as ifg, sum(innouts) as ifo, sum(a) as ifa, sum(e) as ife
FROM
	lahman2017.fielding_summary_final
where pos in ('1b', '2b', '3b', 'ss')
group by playerid;

drop table if exists t_all_if_fielding_summary;

create temporary table t_all_if_fielding_summary as 
select
	fielding_summary_final.playerid, fielding_summary_final.g as ag, fielding_summary_final.innouts as ao,
		fielding_summary_final.a as aa, fielding_summary_final.e as ae,
	t_if_fielding_summary_final.ifg as ifg,
    t_if_fielding_summary_final.ifo as ifo,
    t_if_fielding_summary_final.ifa as ifa,
	t_if_fielding_summary_final.ife as ife
from
	fielding_summary_final left join t_if_fielding_summary_final
on
	fielding_summary_final.playerid =  t_if_fielding_summary_final.playerid
     where not fielding_summary_final.pos in ('1b', '2b', '3b', 'ss');
    
drop table if exists t_all_if_of_fielding_summary;

create temporary table t_all_if_of_fielding_summary as 
select
	t_all_if_fielding_summary.*, b.ofg, b.ofo, b.ofa, b.ofe
    from 
    t_all_if_fielding_summary
    left join
		(select playerid, g as ofg, innouts as ofo, a as ofa, e as ofe from fielding_summary_final where pos='OF') as b
	on
    t_all_if_fielding_summary.playerid=b.playerid;
    
drop table if exists t_all_if_of_p_fielding_summary;

create temporary table t_all_if_of_p_fielding_summary as 
select
	t_all_if_of_fielding_summary.*, b.pg, b.po, b.pa, b.pe
    from 
    t_all_if_of_fielding_summary
    left join
		(select playerid, g as pg, innouts as po, a as pa, e as pe from fielding_summary_final where pos='P') as b
	on
    t_all_if_of_fielding_summary.playerid=b.playerid;
    
drop table if exists t_all_if_of_p_c_fielding_summary; 

create temporary table t_all_if_of_p_c_fielding_summary as 
select
	t_all_if_of_p_fielding_summary.*, b.cg, b.co, b.ca, b.ce
    from 
    t_all_if_of_p_fielding_summary
    left join
		(select playerid, g as cg, innouts as co, a as ca, e as ce from fielding_summary_final where pos='C') as b
	on
    t_all_if_of_p_fielding_summary.playerid=b.playerid;
    
 drop table if exists  fielding_summary_all_final;  
create table fielding_summary_all_final as select * from t_all_if_of_p_c_fielding_summary;

drop table if exists b_p_summary;

create temporary table b_p_summary as select
	batting_summary_final.playerid,
    batting_summary_final.g as bg, batting_summary_final.ab, batting_summary_final.r,
		batting_summary_final.h, batting_summary_final.b2, batting_summary_final.b3,
        batting_summary_final.hr, batting_summary_final.rbi, batting_summary_final.bb,
        batting_summary_final.ibb,
	pitching_summary_final.g as pitch_g, pitching_summary_final.g as pitch_w,
		pitching_summary_final.l as pitch_l, pitching_summary_final.ipouts as pitch_ipo,
        pitching_summary_final.h as pitch_h, pitching_summary_final.er as pitch_er
from
	batting_summary_final left join pitching_summary_final
on
	batting_summary_final.playerid=pitching_summary_final.playerid
union
select
	pitching_summary_final.playerid,
    batting_summary_final.g as bg, batting_summary_final.ab, batting_summary_final.r,
		batting_summary_final.h, batting_summary_final.b2, batting_summary_final.b3,
        batting_summary_final.hr, batting_summary_final.rbi, batting_summary_final.bb,
        batting_summary_final.ibb,
	pitching_summary_final.g as pitch_g, pitching_summary_final.g as pitch_w,
		pitching_summary_final.l as pitch_l, pitching_summary_final.ipouts as pitch_ipo,
        pitching_summary_final.h as pitch_h, pitching_summary_final.er as pitch_er
from
	batting_summary_final right join pitching_summary_final
on
	batting_summary_final.playerid=pitching_summary_final.playerid;
    
create table x_b_p_summary as select * from b_p_summary;

drop table if exists b_p_f_summary;

create temporary table b_p_f_summary as 
select
	x_b_p_summary.playerid,
    x_b_p_summary.bg as bg, x_b_p_summary.ab, x_b_p_summary.r,
		x_b_p_summary.h, x_b_p_summary.b2, x_b_p_summary.b3,
        x_b_p_summary.hr, x_b_p_summary.rbi, x_b_p_summary.bb,
        x_b_p_summary.ibb,
        x_b_p_summary.pitch_g, x_b_p_summary.pitch_w, x_b_p_summary.pitch_l,
        x_b_p_summary.pitch_ipo, x_b_p_summary.pitch_h, x_b_p_summary.pitch_er,
	fielding_summary_all_final.ag, fielding_summary_all_final.ao, fielding_summary_all_final.aa, fielding_summary_all_final.ae,
    fielding_summary_all_final.ifg, fielding_summary_all_final.ifo, fielding_summary_all_final.ifa, fielding_summary_all_final.ife,
    fielding_summary_all_final.ofg, fielding_summary_all_final.ofo, fielding_summary_all_final.ofa, fielding_summary_all_final.ofe,
    fielding_summary_all_final.pg, fielding_summary_all_final.po, fielding_summary_all_final.pa, fielding_summary_all_final.pe,
    fielding_summary_all_final.cg, fielding_summary_all_final.co, fielding_summary_all_final.ca, fielding_summary_all_final.ce
from
	x_b_p_summary left join fielding_summary_all_final
on
	x_b_p_summary.playerid=fielding_summary_all_final.playerid
union
select
	fielding_summary_all_final.playerid,
    x_b_p_summary.bg as bg, x_b_p_summary.ab, x_b_p_summary.r,
		x_b_p_summary.h, x_b_p_summary.b2, x_b_p_summary.b3,
        x_b_p_summary.hr, x_b_p_summary.rbi, x_b_p_summary.bb,
        x_b_p_summary.ibb,
        x_b_p_summary.pitch_g, x_b_p_summary.pitch_w, x_b_p_summary.pitch_l,
        x_b_p_summary.pitch_ipo, x_b_p_summary.pitch_h, x_b_p_summary.pitch_er,
	fielding_summary_all_final.ag, fielding_summary_all_final.ao, fielding_summary_all_final.aa, fielding_summary_all_final.ae,
    fielding_summary_all_final.ifg, fielding_summary_all_final.ifo, fielding_summary_all_final.ifa, fielding_summary_all_final.ife,
    fielding_summary_all_final.ofg, fielding_summary_all_final.ofo, fielding_summary_all_final.ofa, fielding_summary_all_final.ofe,
    fielding_summary_all_final.pg, fielding_summary_all_final.po, fielding_summary_all_final.pa, fielding_summary_all_final.pe,
    fielding_summary_all_final.cg, fielding_summary_all_final.co, fielding_summary_all_final.ca, fielding_summary_all_final.ce
from
	x_b_p_summary right join fielding_summary_all_final
on
	x_b_p_summary.playerid=fielding_summary_all_final.playerid;
    
create table player_performance_summary_final as select * from b_p_f_summary;

create table batting_pitching_final as select * from x_b_p_summary;

```

In [149]:
%sql SELECT * FROM lahman2017.batting_pitching_final order by playerid limit 10;

 * mysql+pymysql://dbuser:***@localhost/lahman2017
10 rows affected.


[('aardsda01', None, None, None, None, None, None, None, None, None, None, 129.0, 129.0, 9.0, 360.0, 97.0, 54.0),
 ('abadfe01', None, None, None, None, None, None, None, None, None, None, 345.0, 345.0, 25.0, 915.0, 287.0, 120.0),
 ('abreubo01', 466.0, 1403.0, 182.0, 354.0, 88.0, 3.0, 32.0, 171.0, 220.0, 11.0, None, None, None, None, None, None),
 ('abreujo02', 614.0, 2414.0, 330.0, 726.0, 144.0, 12.0, 124.0, 410.0, 172.0, 39.0, None, None, None, None, None, None),
 ('accarje01', None, None, None, None, None, None, None, None, None, None, 57.0, 57.0, 3.0, 219.0, 81.0, 42.0),
 ('aceveal01', None, None, None, None, None, None, None, None, None, None, 145.0, 145.0, 15.0, 763.0, 229.0, 117.0),
 ('ackledu01', 584.0, 2012.0, 249.0, 488.0, 91.0, 16.0, 42.0, 201.0, 182.0, 10.0, None, None, None, None, None, None),
 ('acostma01', None, None, None, None, None, None, None, None, None, None, 130.0, 130.0, 6.0, 402.0, 128.0, 65.0),
 ('adamsau01', None, None, None, None, None, None, None, None, None, None, 47.0, 47.0, 0.0, 155.0, 64.0, 34.0),
 ('adamsma01', 528.0, 1586.0, 194.0, 431.0, 95.0, 6.0, 72.0, 255.0, 103.0, 12.0, None, None, None, None, None, None)]

__Your Simplified Answer__

```
%%sql create table batting_summary_final as 
    (select playerid, sum(g) as g, sum(ab) as ab, sum(r) as r, sum(h) as h, sum(`2b`) as b2, sum(`3b`) as b3,
    sum(hr) as hr, sum(rbi) as rbi, sum(bb) as bb, sum(ibb) as ibb 
    FROM lahman2017.batting where yearid >= 2010 and ab >= 100 
    group by playerid
    having ab >= 500);

%%sql create table pitching_summary_final as SELECT
    playerid, sum(g) as g, sum(w) as w, sum(l) as l, sum(IPOuts) as ipouts, sum(h) as h, sum(er) as er
    FROM lahman2017.Pitching
    where yearid >= 2010 and ipouts >= 50
    group by playerid
    having sum(ipouts) >= 150;
    
%%sql create temporary table b_p_summary as select
        batting_summary_final.playerid,
        batting_summary_final.g as bg, batting_summary_final.ab, batting_summary_final.r,
            batting_summary_final.h, batting_summary_final.b2, batting_summary_final.b3,
            batting_summary_final.hr, batting_summary_final.rbi, batting_summary_final.bb,
            batting_summary_final.ibb,
        pitching_summary_final.g as pitch_g, pitching_summary_final.g as pitch_w,
            pitching_summary_final.l as pitch_l, pitching_summary_final.ipouts as pitch_ipo,
            pitching_summary_final.h as pitch_h, pitching_summary_final.er as pitch_er
    from
        batting_summary_final left join pitching_summary_final
    on
        batting_summary_final.playerid=pitching_summary_final.playerid
    union
    select
        pitching_summary_final.playerid,
        batting_summary_final.g as bg, batting_summary_final.ab, batting_summary_final.r,
            batting_summary_final.h, batting_summary_final.b2, batting_summary_final.b3,
            batting_summary_final.hr, batting_summary_final.rbi, batting_summary_final.bb,
            batting_summary_final.ibb,
        pitching_summary_final.g as pitch_g, pitching_summary_final.g as pitch_w,
            pitching_summary_final.l as pitch_l, pitching_summary_final.ipouts as pitch_ipo,
            pitching_summary_final.h as pitch_h, pitching_summary_final.er as pitch_er
    from
        batting_summary_final right join pitching_summary_final
    on
        batting_summary_final.playerid=pitching_summary_final.playerid;

%%sql create table x_b_p_summary as select * from b_p_summary;

%%sql create table batting_pitching_final as select * from x_b_p_summary;
```

## Schema Refinement and Improvement (Question 3.4)

### Statement

- The table definitions for classiccars.orderdetails and classiccars.products are below.


- For the purposes of this question, you do not have to consider any other tables or relationships between the tables below and other tables in the schema/database.

```
CREATE TABLE `orderdetails` (
  `orderNumber` int(11) NOT NULL,
  `productCode` varchar(15) NOT NULL,
  `quantityOrdered` int(11) NOT NULL,
  `priceEach` decimal(10,2) NOT NULL,
  `orderLineNumber` smallint(6) NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

CREATE TABLE `products` (
  `productCode` varchar(15) NOT NULL,
  `productName` varchar(70) NOT NULL,
  `productLine` varchar(50) NOT NULL,
  `productScale` varchar(10) NOT NULL,
  `productVendor` varchar(50) NOT NULL,
  `productDescription` text NOT NULL,
  `quantityInStock` smallint(6) NOT NULL,
  `buyPrice` decimal(10,2) NOT NULL,
  `MSRP` decimal(10,2) NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

- There are several design flaws with the two tables.  Probing will reveal some possible improvements.


- Produce an improved, cleaned up schema for the two tables below. Your answer probably should include triggers. You only need to consider INSERT if you implement triggers.


- Do not ask how me how many improvements I want you to make. If you have a job and your manager asks you to fix a defective schema, I suggest that you do not respond, "How many improvements do you want me to make?"

### Answer

Put the created table statements for the modified schema below. You do not need to execute the statements.


```
CREATE TABLE `orderdetails` (
  `orderNumber` int(11) NOT NULL,
  `productCode` varchar(15) NOT NULL,
  `quantityOrdered` int(11) NOT NULL,
  `priceEach` decimal(10,2) NOT NULL,
  `orderLineNumber` smallint(6) NOT NULL,
  PRIMARY KEY(orderNumber)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

DELIMITER$
create trigger orderdetails_trigger
before insert on orderdetails for each row
begin
    if new.orderNumber <= 0, then
    SIGNAL SQLSTATE '45001'
			SET MESSAGE_TEXT = 'orderNumber invalid', MYSQL_ERRNO=9001;
    end if;      
    if new.quantityOrdered < 0, then
    SIGNAL SQLSTATE '45001'
			SET MESSAGE_TEXT = 'quantityOrdered invalid', MYSQL_ERRNO=9001;
    end if;
    if new.priceEach < 0, then
    SIGNAL SQLSTATE '45001'
			SET MESSAGE_TEXT = 'priceEach invalid', MYSQL_ERRNO=9001;
    end if;
    if new.orderLineNumber <= 0, then
    SIGNAL SQLSTATE '45001'
			SET MESSAGE_TEXT = 'orderNumber invalid', MYSQL_ERRNO=9001;
    end if;      
end$
DELIMITER;

CREATE TABLE `products` (
  `productCode` varchar(15) NOT NULL,
  `productName` varchar(70) NOT NULL,
  `productLine` varchar(50) NOT NULL,
  `productScale` varchar(10) NOT NULL,
  `productVendor` varchar(50) NOT NULL,
  `productDescription` text NOT NULL,
  `quantityInStock` smallint(6) NOT NULL,
  `buyPrice` decimal(10,2) NOT NULL,
  `MSRP` decimal(10,2) NOT NULL,
  PRIMARY KEY(productCode)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

DELIMITER$
create trigger products_trigger
before insert on products for each row
begin
    if new.quantityInStock < 0, then
    SIGNAL SQLSTATE '45001'
			SET MESSAGE_TEXT = 'quantityInStock invalid', MYSQL_ERRNO=9001;
    end if;      
    if new.buyPrice < 0, then
    SIGNAL SQLSTATE '45001'
			SET MESSAGE_TEXT = 'buyPrice invalid', MYSQL_ERRNO=9001;
    end if;
    if new.MSRP < 0, then
    SIGNAL SQLSTATE '45001'
			SET MESSAGE_TEXT = 'MSRP invalid', MYSQL_ERRNO=9001;
    end if;    
end$
DELIMITER;
```

## SQL Modeling $-$ Inheritance (Question 3.5)

### Statement

- A _Person_ has the following properties:
    - _UNI_
    - $last\_name$
    - $first\_name$
    - _email_
    
    
- A _Student_ _isA_ _Person_ and has the additional property _major._


- A _Faculty_ _isA_ _Person_ and has the additional property _title._


- No one is both a _Student_ and _Faculty._ All instances of _Person_ are either a _Student_ or a _Faculty._


- A _Student_ has exactly one _Faculty_ advisor.


- A _Faculty_ advised zero or more _Students._


- Define an SQL Schema for representing _Person, Student_, _Faculty_ and the _Advises_ relationship. Draw a simple diagram.


- You must write several functions (in Python, Java, etc) that use SQL.
    1. create_student(): Computes a unique UNI based on the first two letters of the first and last name and a count of the number of entries that have the same two characters for first and last names. The function also saves the properties in one or more tables.
    2. retrieve_person(uni) returns the corresponding _Person_ information.
    3. retrieve_student(uni) returns the corresponding _Student_ information.
    
    
- $create\_student()$ may require a transaction. If you think it does, implement create using a transaction.


- Do not use triggers or stored procedures.

### Answer

#### DDL/Schema


![image.png](3.5.png)
```
CREATE TABLE `person` (
  `uni` varchar(12) NOT NULL,
  `last_name` varchar(45) NOT NULL,
  `first_name` varchar(45) NOT NULL,
  `email` varchar(45) NOT NULL,
  PRIMARY KEY (`uni`),
  UNIQUE KEY `email_UNIQUE` (`email`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

CREATE TABLE `student` (
  `uni` varchar(12) NOT NULL,
  `major` varchar(12) NOT NULL,
  `advisor` varchar(12) NOT NULL,
  PRIMARY KEY (`uni`),
  KEY `student_faculty_idx` (`advisor`),
  CONSTRAINT `student_faculty` FOREIGN KEY (`advisor`) REFERENCES `faculty` (`uni`) 
      ON DELETE NO ACTION ON UPDATE NO ACTION,
  CONSTRAINT `student_person` FOREIGN KEY (`uni`) REFERENCES `person` (`uni`) 
      ON DELETE NO ACTION ON UPDATE NO ACTION
) ENGINE=InnoDB DEFAULT CHARSET=utf8;


CREATE TABLE `faculty` (
  `uni` varchar(12) NOT NULL,
  `title` varchar(45) NOT NULL,
  PRIMARY KEY (`uni`),
  CONSTRAINT `faculty_person` FOREIGN KEY (`uni`) REFERENCES `person` (`uni`) ON DELETE NO ACTION ON UPDATE NO ACTION
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

```

#### Functions

In [37]:
import pymysql
import json
import pymysql.cursors

def compute_uni_prefix(last_name, first_name):
    result = last_name[0:2] + first_name[0:2]
    return result

def create_faculty(last_name, first_name, email, title):
    try:
        params['dbname'] = "w4111final"
        cnx = get_new_connection(params)
        cur = cnx.cursor()
        cur.execute("SET SESSION TRANSACTION ISOLATION LEVEL SERIALIZABLE")
        
        prefix = compute_uni_prefix(last_name, first_name)
        prefix = prefix.lower()
        
        q1 = "select count(*) as count from person where uni like (%s)"
        result = run_q(cnx, q1, prefix + "%", True, False)
        
        count = result[0]['count'] + 1
        new_uni = prefix + str(count)
        
#         print(new_uni)
        
        q2 = "insert into person values (%s, %s, %s, %s)"
        r2 = run_q(cnx, q2, (new_uni, last_name, first_name, email), commit = True)
        
#         print(r2)
        
        q3 = "insert into faculty values (%s, %s)"
        r3 = run_q(cnx, q3, (new_uni, title), commit = True)
        
#         print(r3)
        
        cnx.commit()
        cnx.close()
        
    except Exception as e:
        print("Exception =", e)
        cnx.rollback()
        cnx.close()
        
def create_student(last_name, first_name, email, major, advisor_uni):
    try:
        params['dbname'] = "w4111final"
        cnx = get_new_connection(params)
        cur = cnx.cursor()
        cur.execute("SET SESSION TRANSACTION ISOLATION LEVEL SERIALIZABLE")
        
        prefix = compute_uni_prefix(last_name, first_name)
        prefix = prefix.lower()
        
        q1 = "select count(*) as count from person where uni like (%s)"
        result = run_q(cnx, q1, prefix + "%", True, False)
        
        count = result[0]['count'] + 1
        new_uni = prefix + str(count)
        
        q2 = "insert into person values (%s, %s, %s, %s)"
        r2 = run_q(cnx, q2, (new_uni, last_name, first_name, email), commit = True)
        
        q3 = "insert into student values (%s, %s, %s)"
        r3 = run_q(cnx, q3, (new_uni, major, advisor_uni), commit = True)
        
        cnx.commit()
        cnx.close()
        
    except Exception as e:
        print("Exception =", e)
        cnx.rollback()
        cnx.close()
        
def retrieve_person(uni):
    try:
        params['dbname'] = "w4111final"
        cnx = get_new_connection(params)
        cur = cnx.cursor()
        
        q1 = "select * from person where uni = %s"
        result = run_q(cnx, q1, uni, True, True)
        
        result = result[0]
        cnx.close()
        return result
    
    except Exception as e:
        print("Exception =", e)
        cnx.close()
        
def retrieve_student(uni):
    try:
        params['dbname'] = "w4111final"
        cnx = get_new_connection(params)
        cur = cnx.cursor()
        
        q1 = "select person.*, student.major, student.advisor from person join student on \
                student.uni = person.uni where person.uni = %s"
        result = run_q(cnx, q1, uni, True, True)
        
        result = result[0]
        cnx.close()
        return result
    
    except Exception as e:
        print("Exception =", e)
        cnx.close()

#### Tests

In [36]:
create_faculty("Wizard", "Sauron", "darklord@columbia.edu", "Professor")

In [38]:
p = retrieve_person('wisa1')
print("Person = ", json.dumps(p, indent=2))

Person =  {
  "uni": "wisa1",
  "last_name": "Wizard",
  "first_name": "Sauron",
  "email": "darklord@columbia.edu"
}


In [39]:
create_student("Gollum", "Unknown", "collum@columbia.edu", "theft", "wisa1")

In [40]:
p = retrieve_student('goun1')
print("Person = ", json.dumps(p, indent=2))

Person =  {
  "uni": "goun1",
  "last_name": "Gollum",
  "first_name": "Unknown",
  "email": "collum@columbia.edu",
  "major": "theft",
  "advisor": "wisa1"
}


In [41]:
create_student("Gollumer", "Unknown", "collumer@columbia.edu", "theft", "wisa1")

In [42]:
p = retrieve_student('goun2')
print("Person = ", json.dumps(p, indent=2))

Person =  {
  "uni": "goun2",
  "last_name": "Gollumer",
  "first_name": "Unknown",
  "email": "collumer@columbia.edu",
  "major": "theft",
  "advisor": "wisa1"
}


In [43]:
create_student("Gollumer", "Unknown", "collumer@columbia.edu", "theft", "wisal")

Exception = (1062, "Duplicate entry 'collumer@columbia.edu' for key 'email_UNIQUE'")


In [44]:
create_student("Gollumor", "Unknown", "collumor@columbia.edu", "theft", "1111")

Exception = (1452, 'Cannot add or update a child row: a foreign key constraint fails (`w4111final`.`student`, CONSTRAINT `student_faculty` FOREIGN KEY (`advisor`) REFERENCES `faculty` (`uni`))')


## Graph Database $-$ Neo4j (Question 3.6)

### Statement

- The problem statement is exactly the same as for  3.5 SQL Modeling $-$ Inheritance.


- Draw a diagram of your data model.


- Implement the same functions. You DO NOT have to worry about transactions.

### Answer


#### Diagram



![image.png](361.png)
![image.png](362.png)
![image.png](363.png)
![image.png](364.png)

#### Functions

In [112]:
from py2neo import Graph, NodeMatcher, Node, RelationshipMatcher, Relationship
import json
import uuid
import copy

class FanGraph(object):

    # Connects to the DB and sets a Graph instance variable.
    # Also creates a NodeMatcher, which is a py2neo class.
    def __init__(self, auth, host, port, secure=False, ):
        self._graph = Graph(secure=secure,
                            bolt=True,
                            auth=auth,
                            host=host,
                            port=port)
        self._node_matcher = NodeMatcher(self._graph)
        self._relationship_matcher = RelationshipMatcher(self._graph)
        
    def run_match(self, labels=None, properties=None):
        """
        Uses a NodeMatcher to find a node matching a "template."
        :param labels: A list of labels that the node must have.
        :param properties: A parameter list of the form prop1=value1, prop2=value2, ...
        :return: An array of Node objects matching the pattern.
        """
        #ut.debug_message("Labels = ", labels)
        #ut.debug_message("Properties = ", json.dumps(properties))

        if labels is not None and properties is not None:
            result = self._node_matcher.match(labels, **properties)
        elif labels is not None and properties is None:
            result = self._node_matcher.match(labels)
        elif labels is None and properties is not None:
            result = self._node_matcher.match(**properties)
        else:
            raise ValueError("Invalid request. Labels and properties cannot both be None.")

        # Convert NodeMatch data into a simple list of Nodes.
        full_result = []
        for r in result:
            full_result.append(r)

        return full_result
        
    def find_nodes_by_template(self, tmp):
        """

        :param tmp: A template defining the label and properties for Nodes to return. An
         example is { "label": "Fan", "template" { "last_name": "Ferguson", "first_name": "Donald" }}
        :return: A list of Nodes matching the template.
        """
        labels = tmp.get('label')
        props = tmp.get("template")
        result = self.run_match(labels=labels, properties=props)
        return result

    def compute_uni_prefix(last_name, first_name):
        result = last_name[0:2] + first_name[0:2]
        print(result)
        return result

    def create_faculty(self, last_name, first_name, email, title):
        prefix = compute_uni_prefix(last_name, first_name)
        prefix = prefix.lower()

        tx = self._graph.begin(autocommit = False)

        try:
            q1 = "match (p:person) where p.uni starts with {x} return count(p) as count"
            cnt = self._graph.run(q1, x = prefix)
            cnt = cnt.data()
            cnt = cnt[0]['count'] + 1

            new_uni = prefix + str(cnt)

            n = Node("person", "faculty", uni = new_uni, last_name = last_name, first_name = first_name, email = email, title = title)

            tx.create(n)
            tx.commit()
            return new_uni

        except Exception as e:
            print("Exception =", e)
            tx.rollback()

    def create_student(self, last_name, first_name, email, major, advisor):
        prefix = compute_uni_prefix(last_name, first_name)
        prefix = prefix.lower()

        tx = self._graph.begin(autocommit = False)

        try:
            q1 = "match (p:person) where p.uni starts with {x} return count(p) as count"
            cnt = self._graph.run(q1, x = prefix)
            cnt = cnt.data()
            cnt = cnt[0]['count'] + 1

            new_uni = prefix + str(cnt)

            f = self.get_faculty(advisor)
            if f is None:
                raise ValueError("Advisor doesn't exist")

            n = Node("person", "student", uni = new_uni, last_name = last_name, first_name = first_name, email = email, major = major, advisor = advisor)       
            tx.create(n)

            r = Relationship(f, "advises", n)
            tx.create(r)

            tx.commit()
            return new_uni

        except Exception as e:
            print("Exception =", e)
            tx.rollback()

    def get_faculty(self, uni):
        n = self.find_nodes_by_template({"label":"faculty", "template":{"uni":uni}})
        if n is not None and len(n) > 0:
            n = n[0]
        else:
            n = None
        return n

    def get_person(self, uni):
        person_field = ['uni', 'last_name', 'first_name', 'email']
        n = self.find_nodes_by_template({"label":"person", "template":{"uni":uni}})
        if n is not None and len(n) > 0:
            n = n[0]
        else:
            n = None
        return n

        key = copy.copy(list(n.keys()))
        for k in key:
            if k not in person_field:
                del(n[k])
        return n

    def get_student(self, uni):
        n = self.find_nodes_by_template({"label":"student", "template":{"uni":uni}})
        if n is not None and len(n) > 0:
            n = n[0]
        else:
            n = None
        return n

#### Tests

In [113]:
fg = FanGraph(auth=('dbuser','dbuser'),
                              host="localhost",
                              port=7687,
                              secure=False)


uni = fg.create_faculty("Ferguson", "Donald", "dff@cs.columbia.edu", "Professor")
print("Faculty uni = ", uni)
uni = fg.create_student("Baggins", "Bilbo", "bb@cs.columbia.edu", "Magic", "fedo1")
result = fg.get_person(uni)
print('Person = ', result)
result = fg.get_student(uni)
print('Student = ', result)

Faculty uni =  fedo1
Person =  (_19507:person:student {advisor: 'fedo1', email: 'bb@cs.columbia.edu', first_name: 'Bilbo', last_name: 'Baggins', major: 'Magic', uni: 'babi1'})
Student =  (_19507:person:student {advisor: 'fedo1', email: 'bb@cs.columbia.edu', first_name: 'Bilbo', last_name: 'Baggins', major: 'Magic', uni: 'babi1'})


In [114]:
print('Student = ', fg.get_student('1111'))

Student =  None


## Key/Value Modeling $-$ Redis (Question 3.7)

### Statement

- You guessed it. Do the same for Redis.

### Answer

#### Diagram



#### Functions

![image.png](37.png)

In [92]:
import redis
from operator import itemgetter
import json
import copy

"""
Connect to local Redis server. StrictRedis complies more closely with standard than
simple Redis client in this package. decode_responses specifies whether or not to convert
bytes to UTF8 character string or treat as raw bytes, e.g. an image, audio stream, etc.
"""
r = redis.StrictRedis(
    host='localhost',
    port=6379,
    charset="utf-8", decode_responses=True)


def add_to_cache(key, value):
    """

    :param key: A valid Redis key string.
    :param value: A Python dictionary to add to cache.
    :return: None
    """
    r.hmset(key, value)


def get_from_cache(key):
    """

    :param key: A valid Redis key.
    :return: The "map object" associated with the key.
    """
    result = r.hgetall(key)

    return result


def get_uni(last_name, first_name):

    uni_prefix = last_name[0:2] + first_name[0:2]
    uni_prefix = uni_prefix.lower()

    key_template = "person:*:" + uni_prefix + "*"
    result = r.keys(key_template)
    next_num = len(result)+1

    result = uni_prefix + str(next_num)

    return result


def create_faculty(last_name, first_name, email, titles):
    new_uni = get_uni(last_name, first_name)
    new_key = "person:faculty:" + new_uni
    props = {
        "last_name": last_name,
        "first_name": first_name,
        "email": email,
        "titles": titles,
        "uni": new_uni
    }
    add_to_cache(new_key, props)
    return new_uni

def create_student(last_name, first_name, email, major, advisor):
    new_uni = get_uni(last_name, first_name)
    new_key = "person:student:" + new_uni
    props = {
        "last_name": last_name,
        "first_name": first_name,
        "email": email,
        "major": major,
        "uni": new_uni
    }
    add_to_cache(new_key, props)
    r.lpush("Advisors:" + advisor, new_uni)
    return new_uni

def get_student(uni):
    key = "person:student:" + uni
    result = get_from_cache(key)
    return result

def get_person(uni):
    person_fields = ['uni', 'last_name', 'first_name', 'email']
    partten = "person:*:" + uni
    keys = r.keys(partten)
    
    result = None
    
    if keys is not None and len(keys) == 1:
        k = keys[0]
        tmp = get_from_cache(k)
        if tmp is not None and len(tmp) > 0:
            result = tmp
        if k is not None:
            keys = copy.copy(list(result.keys()))
            for key in keys:
                if key not in person_fields:
                    del(result[key])
    return result

def get_advises(uni):
    key = "Advisors:" + uni
    result = r.lrange(key, 0, -1)
    return result

### Tests

In [93]:
u1 = create_student("Poppins", "Mary", "mp@cs.columbia.edu", "Magic", funi)
print(u1)

poma7


In [94]:
u2 = create_student("Ponce", "Mary", "mp@cs.columbia.edu", "Magic", funi)
print(u2)

poma8


In [80]:
funi = create_faculty("Ferguson", "Donald", "dff@cs.columbia.edu", "Professor")
print(funi)

fedo3


In [115]:
funi = create_faculty("Ferguson", "Donald", "dff@cs.columbia.edu", "Professor")
u1 = create_student("Poppins", "Mary", "mp@cs.columbia.edu", "Magic", funi)
u2 = create_student("Ponce", "Mary", "mp@cs.columbia.edu", "Magic", funi)

print("Person 1 = ", get_person(u1))
print("Student 2 =  ", get_student(u2))

print("Advisees = ", get_advises(funi))

Person 1 =  {'last_name': 'Poppins', 'first_name': 'Mary', 'email': 'mp@cs.columbia.edu', 'uni': 'poma9'}
Student 2 =   {'last_name': 'Ponce', 'first_name': 'Mary', 'email': 'mp@cs.columbia.edu', 'major': 'Magic', 'uni': 'poma10'}
Advisees =  ['poma10', 'poma9']


In [117]:
print("Student 2 =  ", get_student('aaaa'))

Student 2 =   {}


## Transactions (Question 3.8)

### Statement

- Below is a table definition for a banking account.

```
CREATE TABLE `banking_account` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `balance` double NOT NULL,
  `version` varchar(64) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=utf8;
```

- The fields have the following meanings:
    - _id_ is the account number or ID. The field is auto-increment, which means a new ID is created om every INSERT.
    - _balance_ is the account balance.
    - _version_ is a UUID in string format. Every program that updates an account tuple generates a new UUID and updates on the insert along with any other fields the program changes.
    

- Write two programs:
    - $transfer\_pessimistic:$
        - Begins a transaction and sets the isolation level to serializable.
        - Prompts the user for the source account ID.
        - Reads and displays the balance and asks the user if they want to continue.
        - Prompts the user for the target account ID.
        - Reads and displays the balance and asks the user if they want to continue.
        - Prompts the user for the amount to transfer.
        - Performs the transfer.
        - Commits.
    - $transfer\_optimistic:$
        - Prompts the user for the source account ID.
        - Reads and displays the balance and asks the user if they want to continue.
        - Prompts the user for the target account ID.
        - Reads and displays the balance and asks the user if they want to continue.
        - Prompts the user for the amount to transfer.
        - Begins a transaction and sets the isolation level to serializable.
        - Rereads the information.
        - If the version data has not changed, performs the transfer.
        - Commits.
        

### Answer

#### Functions

In [118]:
def create_account(balance):
    params['dbname'] = "w4111final"
    cnx = get_new_connection(params)
    cur = cnx.cursor()
    cur.execute("SET SESSION TRANSACTION ISOLATION LEVEL SERIALIZABLE")
    
    version_id = str(uuid.uuid4())
    q = "insert into banking_account (balance, version) values (%s, %s)"
    result = run_q(cnx, q, (balance, version_id), True, commit = True)
    
    q2 = "select max(id) as new_id from banking_account;"
    result2 = run_q(cnx, q2, None, fetch = True, commit = False)
    
    result2 = result2[0]['new_id']
    cnx.commit()
    cnx.close()
    
    return result2

In [100]:
def get_balance(id, cursor=None):
    """
    Same logic as above. Normally, there would be a single function that returned data based on
    requested fields instead of two different functions.
    """
    
    if cursor is None:
        params['dbname'] = "w4111final"
        cnx = get_new_connection(params)
        cur = cnx.cursor()
        cursor_created = True
    else:
        cursor_created = False
        cnx = None

    q = "select * from banking_account where id=%s"
    result = run_q(cnx, q, id, fetch=True, commit=False)

    if cursor_created:
        cnx.commit()
        cnx.close()

    return result[0]['balance']

In [101]:
def get_account(id, cursor=None):
    """
    Same logic as above. Normally, there would be a single function that returned data based on
    requested fields instead of two different functions.
    """
    
    if cursor is None:
        params['dbname'] = "w4111final"
        cnx = get_new_connection(params)
        cur = cnx.cursor()
        cursor_created = True
    else:
        cursor_created = False
        cnx = None

    q = "select * from banking_account where id=%s"
    result = run_q(cnx, q, id, fetch=True, commit=False)

    if cursor_created:
        cnx.commit()
        cnx.close()

    return result[0]

In [123]:
def update_balance(id, amount, cursor = None):
    cnx = None
    if cursor is None:
        params['dbname'] = "w4111final"
        cnx = get_new_connection(params)
        cur = cnx.cursor()
        cur.execute("SET SESSION TRANSACTION ISOLATION LEVEL SERIALIZABLE")
        cursor_created = True
    else:
        cursor_created = False
        
    new_version = str(uuid.uuid4())
    q = "update banking_account set balance = %s, version = %s where id = %s"
    result = run_q(cnx, q, (amount, new_version, id), fetch = True, commit = True, cursor = cursor)
    
    if cursor_created:
        cnx.commit()
        cnx.close()

In [103]:
def update_balance_optimistic(acct, amount, cursor = None):
    cnx = None
    if cursor is None:
        params['dbname'] = "w4111final"
        cnx = get_new_connection(params)
        cur = cnx.cursor()
        cur.execute("SET SESSION TRANSACTION ISOLATION LEVEL SERIALIZABLE")
        cursor_created = True
    else:
        cursor_created = False
        
    current_account = get_account(acct['id'])
    if current_account['version'] != acct['version']:
        raise ValueError("Optimistic transaction failed")
        
    new_version = str(uuid.uuid4())
    q = "update banking_account set balance = %s, version = %s where id = %s"
    result = run_q(cnx, q, (amount, new_version, acct['id']), fetch = True, commit = True, cursor = cursor)
    
    if cursor_created:
        cnx.commit()
        cnx.close()

In [104]:
def transfer_pessmistic():
    print("\n*** Transfering Pessmisticlly ***\n")
    params['dbname'] = "w4111final"
    cnx = get_new_connection(params)
    cur = cnx.cursor()
    
    try:
        cur.execute("SET SESSION TRANSACTION ISOLATION LEVEL SERIALIZABLE")
        source_id = input("Source account id: ")
        source_b = get_balance(source_id, cursor = cur)
        cont = input("Source Balance = " + str(source_b) + ". Continue(y/n)")
        
        if cont == "y":
            target_id = input("Target account id: ")
            target_b = get_balance(target_id, cursor = cur)
            cont = input("Target balance = " + str(target_b) + ". Continue(y/n)")
            
            if cont == "y":
                amount = input("Amount: ")
                amount = float(amount)
                new_source = source_b - amount
                new_target = target_b + amount
                
                update_balance(source_id, new_source, cursor = cur)
                update_balance(target_id, new_target, cursor = cur)
                
                cnx.commit()
                cnx.close()
    except Exception as e:
        print("Exception =", e)
        cnx.rollback()
        cnx.close()
        
    return

In [105]:
def transfer_optimistic():
    print("\n*** Transfering optimisticlly ***\n")

    source_id = input("Source account id: ")
    source_acct = get_account(source_id, cursor = None)

    cont = input("Source Balance = " + str(source_acct['balance']) + ". Continue(y/n)")

    if cont == "y":
        target_id = input("Target account id: ")
        target_acct = get_account(target_id, cursor = None)
        cont = input("Target balance = " + str(target_acct['balance']) + ". Continue(y/n)")

        if cont == "y":
            amount = input("Amount: ")
            amount = float(amount)
            new_source = source_acct['balance'] - amount
            new_target = target_acct['balance'] + amount
            
            try:
                params['dbname'] = "w4111final"
                cnx = get_new_connection(params)
                cur = cnx.cursor()
                cur.execute("SET SESSION TRANSACTION ISOLATION LEVEL SERIALIZABLE")

                update_balance_optimistic(source_acct, new_source, cursor = cur)
                update_balance_optimistic(target_acct, new_target, cursor = cur)

                cnx.commit()
                cnx.close()
            except Exception as e:
                print("Exception =", e)
                cnx.rollback()
                cnx.close()
        
    return

#### Tests

In [119]:
id = create_account(300)
print("New account ID = ", id)

New account ID =  7


In [120]:
b = get_balance(id)
print("Account balance is ", b)

Account balance is  300.0


In [121]:
acct = get_account(id)
print("Account = ", json.dumps(acct))

Account =  {"id": 7, "balance": 300.0, "version": "962afbd1-458c-4e08-aef6-2ad145a0ae09"}


In [124]:
update_balance(5, 200)
account = get_account(5)
print("Account = ", json.dumps(account))

Account =  {"id": 5, "balance": 200.0, "version": "e54c1bba-b0b8-491c-ba57-d599d5d5497f"}


_Screenshots of more complex tests._

- At this point, you can run the programs in two different windows to see how they do or do not interfere.


- Screenshots do not really help for all of the tests.


- You can show the optimistic transaction rolling back.


- You can also show deadlocks.

![image.png](3.8.png)

## Star Schema (Question 3.9)

### Statement

Your experience with W4111 have led you to decide that computer science, data science, IEOR, etc. are not the fields for you. While being very talented and good at the domains, the topics bring back back memories. You decide that the world was a better place before computers and you want to student that world. You become an archaeologist.

One day on a dig, you uncover what is an advanced computer inside ancient remains. This is game changing. This is proof of ancient aliens. Despite your aversion to technology, the opportunity is to great and you decide to the student the system. After powering on the device, you discover that it contains a database -- [Sakila Sample Database.] Probing the database reveals that it contains information about video rental stores. Wow! This is really old.

Seeing a killer breakthrough paper, "Uncovered: The Video Viewing Habits of Earth Visiting Ancient Aliens," you decide you must apply data analysis to the data. A snarky voice, sounding suspiciously like the annoying DB professor states, "Use a star schema!"

The two most important facts for the fact table clearly are: a rental, and the corresponding payment. You must build a star schema containing a fact table and dimension tables enabling you to analyze the data.

Your tasks:
1. Define the DDL statements that create the fact table and dimension tables. You must decide what the dimensions are.
2. Write SQL to load/copy data from the base Sakila tables into your fact and dimensions tables.
3. Write three queries demonstrating.
    1. A roll-up operation.
    2. A slice operation.
    3. A dice operaton.
    
    

### Answer


In [129]:
%sql use sakila
%sql select * from rental limit 10

 * mysql+pymysql://dbuser:***@localhost/lahman2017
0 rows affected.
 * mysql+pymysql://dbuser:***@localhost/lahman2017
10 rows affected.


[(1, datetime.datetime(2005, 5, 24, 22, 53, 30), 367, 130, datetime.datetime(2005, 5, 26, 22, 4, 30), 1, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (2, datetime.datetime(2005, 5, 24, 22, 54, 33), 1525, 459, datetime.datetime(2005, 5, 28, 19, 40, 33), 1, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (3, datetime.datetime(2005, 5, 24, 23, 3, 39), 1711, 408, datetime.datetime(2005, 6, 1, 22, 12, 39), 1, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (4, datetime.datetime(2005, 5, 24, 23, 4, 41), 2452, 333, datetime.datetime(2005, 6, 3, 1, 43, 41), 2, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (5, datetime.datetime(2005, 5, 24, 23, 5, 21), 2079, 222, datetime.datetime(2005, 6, 2, 4, 33, 21), 1, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (6, datetime.datetime(2005, 5, 24, 23, 8, 7), 2792, 549, datetime.datetime(2005, 5, 27, 1, 32, 7), 1, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (7, datetime.datetime(2005, 5, 24, 23, 11, 53), 3995, 269, datetime.datetime(2005, 5, 29, 20, 34, 53), 2, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (8, datetime.datetime(2005, 5, 24, 23, 31, 46), 2346, 239, datetime.datetime(2005, 5, 27, 23, 33, 46), 2, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (9, datetime.datetime(2005, 5, 25, 0, 0, 40), 2580, 126, datetime.datetime(2005, 5, 28, 0, 22, 40), 1, datetime.datetime(2006, 2, 15, 21, 30, 53)),
 (10, datetime.datetime(2005, 5, 25, 0, 2, 21), 1824, 399, datetime.datetime(2005, 5, 31, 22, 44, 21), 2, datetime.datetime(2006, 2, 15, 21, 30, 53))]

In [134]:
%%sql
create table rental_payment_facts(
rental_id int(11) not null default'0',
rental_date datetime not null,
inventory_id mediumint(8) unsigned not null,
customer_id smallint(5) unsigned not null,
amount decimal(5,2) not null,
staff_id tinyint(3) unsigned not null
)engine = InnoDB default charset=latin1;

 * mysql+pymysql://dbuser:***@localhost/lahman2017
0 rows affected.


[]

In [135]:
%%sql
create table facts2 as 
(select rental.rental_id, rental.rental_date, rental.inventory_id, rental.customer_id, 
rental.staff_id, payment.payment_id, payment.amount from rental join payment
on rental.rental_id=payment.rental_id)

 * mysql+pymysql://dbuser:***@localhost/lahman2017
16044 rows affected.


[]

In [136]:
%%sql
create temporary table t1 as (
select staff.staff_id, staff.first_name, staff.last_name, store.address_id
from staff join store on staff.store_id=store.store_id)

 * mysql+pymysql://dbuser:***@localhost/lahman2017
2 rows affected.


[]

In [137]:
%%sql
create table data_dimension as (
select rental_date, year(rental_date) as year, ceil(month(rental_date)/4) as quarter, 
month(rental_date) as month, dayofmonth(rental_date) as day from rental);

 * mysql+pymysql://dbuser:***@localhost/lahman2017
16044 rows affected.


[]

In [139]:
%%sql
create table address_dimension as 
select a.*, b.country from (
select address_id, address.city_id, city, district, country_id from address join city
on address.city_id=city.city_id) as a join country as b
on a.country_id=b.country_id

 * mysql+pymysql://dbuser:***@localhost/lahman2017
603 rows affected.


[]

In [141]:
%%sql
create table staff_address_d as (
select t1.*, address_dimension.city, address_dimension.city_id,
address_dimension.district, address_dimension.country, address_dimension.country_id 
from t1 join address_dimension on t1.address_id=address_dimension.address_id);

 * mysql+pymysql://dbuser:***@localhost/lahman2017
2 rows affected.


[]

In [145]:
%%sql
select staff_id, customer_id, sum(amount) from payment 
group by staff_id, customer_id with rollup;

 * mysql+pymysql://dbuser:***@localhost/lahman2017
1201 rows affected.


[(1, 1, Decimal('64.83')),
 (1, 2, Decimal('60.85')),
 (1, 3, Decimal('64.86')),
 (1, 4, Decimal('49.88')),
 (1, 5, Decimal('73.83')),
 (1, 6, Decimal('56.84')),
 (1, 7, Decimal('80.82')),
 (1, 8, Decimal('57.86')),
 (1, 9, Decimal('39.88')),
 (1, 10, Decimal('40.88')),
 (1, 11, Decimal('60.87')),
 (1, 12, Decimal('31.90')),
 (1, 13, Decimal('76.83')),
 (1, 14, Decimal('75.80')),
 (1, 15, Decimal('72.82')),
 (1, 16, Decimal('67.87')),
 (1, 17, Decimal('35.90')),
 (1, 18, Decimal('37.92')),
 (1, 19, Decimal('63.88')),
 (1, 20, Decimal('35.89')),
 (1, 21, Decimal('85.82')),
 (1, 22, Decimal('89.83')),
 (1, 23, Decimal('55.87')),
 (1, 24, Decimal('34.88')),
 (1, 25, Decimal('74.82')),
 (1, 26, Decimal('104.78')),
 (1, 27, Decimal('65.85')),
 (1, 28, Decimal('52.86')),
 (1, 29, Decimal('75.80')),
 (1, 30, Decimal('66.82')),
 (1, 31, Decimal('31.90')),
 (1, 32, Decimal('56.87')),
 (1, 33, Decimal('70.84')),
 (1, 34, Decimal('42.87')),
 (1, 35, Decimal('99.77')),
 (1, 36, Decimal('42.88')),
 (1, 37, Decimal('57.87')),
 (1, 38, Decimal('61.84')),
 (1, 39, Decimal('60.87')),
 (1, 40, Decimal('58.86')),
 (1, 41, Decimal('50.88')),
 (1, 42, Decimal('68.83')),
 (1, 43, Decimal('47.89')),
 (1, 44, Decimal('56.86')),
 (1, 45, Decimal('84.84')),
 (1, 46, Decimal('50.85')),
 (1, 47, Decimal('42.90')),
 (1, 48, Decimal('32.93')),
 (1, 49, Decimal('54.87')),
 (1, 50, Decimal('78.84')),
 (1, 51, Decimal('84.79')),
 (1, 52, Decimal('49.86')),
 (1, 53, Decimal('62.84')),
 (1, 54, Decimal('67.85')),
 (1, 55, Decimal('61.87')),
 (1, 56, Decimal('77.83')),
 (1, 57, Decimal('29.89')),
 (1, 58, Decimal('61.86')),
 (1, 59, Decimal('74.85')),
 (1, 60, Decimal('60.85')),
 (1, 61, Decimal('26.93')),
 (1, 62, Decimal('35.89')),
 (1, 63, Decimal('60.88')),
 (1, 64, Decimal('58.82')),
 (1, 65, Decimal('67.88')),
 (1, 66, Decimal('82.81')),
 (1, 67, Decimal('47.91')),
 (1, 68, Decimal('35.90')),
 (1, 69, Decimal('53.87')),
 (1, 70, Decimal('50.89')),
 (1, 71, Decimal('85.82')),
 (1, 72, Decimal('28.89')),
 (1, 73, Decimal('53.87')),
 (1, 74, Decimal('33.90')),
 (1, 75, Decimal('87.79')),
 (1, 76, Decimal('23.94')),
 (1, 77, Decimal('40.87')),
 (1, 78, Decimal('55.87')),
 (1, 79, Decimal('42.87')),
 (1, 80, Decimal('76.85')),
 (1, 81, Decimal('52.86')),
 (1, 82, Decimal('59.89')),
 (1, 83, Decimal('34.91')),
 (1, 84, Decimal('59.88')),
 (1, 85, Decimal('51.87')),
 (1, 86, Decimal('65.85')),
 (1, 87, Decimal('65.84')),
 (1, 88, Decimal('60.87')),
 (1, 89, Decimal('67.84')),
 (1, 90, Decimal('50.88')),
 (1, 91, Decimal('54.84')),
 (1, 92, Decimal('63.84')),
 (1, 93, Decimal('41.88')),
 (1, 94, Decimal('45.89')),
 (1, 95, Decimal('37.92')),
 (1, 96, Decimal('57.84')),
 (1, 97, Decimal('19.92')),
 (1, 98, Decimal('44.89')),
 (1, 99, Decimal('27.90')),
 (1, 100, Decimal('43.88')),
 (1, 101, Decimal('36.91')),
 (1, 102, Decimal('65.86')),
 (1, 103, Decimal('60.84')),
 (1, 104, Decimal('53.86')),
 (1, 105, Decimal('42.91')),
 (1, 106, Decimal('63.87')),
 (1, 107, Decimal('60.87')),
 (1, 108, Decimal('57.85')),
 (1, 109, Decimal('43.88')),
 (1, 110, Decimal('35.92')),
 (1, 111, Decimal('24.94')),
 (1, 112, Decimal('57.86')),
 (1, 113, Decimal('51.86')),
 (1, 114, Decimal('74.81')),
 (1, 115, Decimal('56.85')),
 (1, 116, Decimal('38.88')),
 (1, 117, Decimal('51.87')),
 (1, 118, Decimal('31.89')),
 (1, 119, Decimal('68.86')),
 (1, 120, Decimal('68.86')),
 (1, 121, Decimal('43.89')),
 (1, 122, Decimal('46.86')),
 (1, 123, Decimal('48.87')),
 (1, 124, Decimal('30.93')),
 (1, 125, Decimal('86.80')),
 (1, 126, Decimal('62.85')),
 (1, 127, Decimal('59.87')),
 (1, 128, Decimal('48.88')),
 (1, 129, Decimal('59.82')),
 (1, 130, Decimal('47.87')),
 (1, 131, Decimal('85.81')),
 (1, 132, Decimal('62.84')),
 (1, 133, Decimal('48.87')),
 (1, 134, Decimal('56.90')),
 (1, 135, Decimal('58.82')),
 (1, 136, Decimal('27.94')),
 (1, 137, Decimal('115.77')),
 (1, 138, Decimal('73.77')),
 (1, 139, Decimal('59.86')),
 (1, 140, Decimal('59.87')),
 (1, 141, Decimal('65.82')),
 (1, 142, Decimal

In [151]:
%%sql
SELECT customer_id,staff_id, amount from payment where customer_id=1

 * mysql+pymysql://dbuser:***@localhost/lahman2017
32 rows affected.


[(1, 1, Decimal('2.99')),
 (1, 1, Decimal('0.99')),
 (1, 1, Decimal('5.99')),
 (1, 2, Decimal('0.99')),
 (1, 2, Decimal('9.99')),
 (1, 1, Decimal('4.99')),
 (1, 1, Decimal('4.99')),
 (1, 2, Decimal('0.99')),
 (1, 1, Decimal('3.99')),
 (1, 2, Decimal('5.99')),
 (1, 1, Decimal('5.99')),
 (1, 1, Decimal('4.99')),
 (1, 1, Decimal('4.99')),
 (1, 1, Decimal('7.99')),
 (1, 2, Decimal('2.99')),
 (1, 1, Decimal('4.99')),
 (1, 2, Decimal('4.99')),
 (1, 1, Decimal('0.99')),
 (1, 2, Decimal('0.99')),
 (1, 2, Decimal('2.99')),
 (1, 2, Decimal('2.99')),
 (1, 2, Decimal('4.99')),
 (1, 2, Decimal('3.99')),
 (1, 1, Decimal('0.99')),
 (1, 2, Decimal('4.99')),
 (1, 1, Decimal('0.99')),
 (1, 2, Decimal('0.99')),
 (1, 2, Decimal('2.99')),
 (1, 1, Decimal('0.99')),
 (1, 1, Decimal('1.99')),
 (1, 2, Decimal('2.99')),
 (1, 1, Decimal('5.99'))]

In [152]:
%%sql
SELECT customer_id,staff_id, amount from payment where staff_id=1 and customer_id=1

 * mysql+pymysql://dbuser:***@localhost/lahman2017
17 rows affected.


[(1, 1, Decimal('2.99')),
 (1, 1, Decimal('0.99')),
 (1, 1, Decimal('5.99')),
 (1, 1, Decimal('4.99')),
 (1, 1, Decimal('4.99')),
 (1, 1, Decimal('3.99')),
 (1, 1, Decimal('5.99')),
 (1, 1, Decimal('4.99')),
 (1, 1, Decimal('4.99')),
 (1, 1, Decimal('7.99')),
 (1, 1, Decimal('4.99')),
 (1, 1, Decimal('0.99')),
 (1, 1, Decimal('0.99')),
 (1, 1, Decimal('0.99')),
 (1, 1, Decimal('0.99')),
 (1, 1, Decimal('1.99')),
 (1, 1, Decimal('5.99'))]

## Graph DB versus Relational DB (Question 3.10)

### Statement

__Note:__ You just have to write the Cypher statement, show result and draw the ER diagram.

1. Use the database from HW4. You can use the Neo4j browser and attach/past the displayed graph in the answer.
2. Write a Cypher query that shows all nodes and relationships with path length two or less starting at nodes that are Fans and ending in nodes that are teams.
3. In your schema W4111, define a data model that would support the Fan, Comment_On, Comment_By, Supports, Follows and Team graph model.
<s>4. Provide the ER diagram and DDL statements for your schema.<s>

<s>5. Load the Fan, Comment_On, Comment_By, Supports, Follows and Team information into the tables you created.</del>
6. Write an SQL query that returns all paths of length two or less starting from a Fan and ending at a Team.
7. Display your query statement and query result.</s>

__Remember that in a graph DB, nodes and relationships do not have a fixed set of properties. You may assume that all property types are strings.__

### Answer

#### Cypher Query and Result



#### SQL Schema



match p = (f2:Fan)-[r*..2]-(t:Team) return p

![image.png](3101.png)

![image.png](3102.png)

#### DDL

#### Loading Code

#### SQL Query

# Written Questions

## Key Compression (Question 4.1)

What is index key compression?  What are two necessary conditions for using index key compression?

_Answer:_

Index key conpression compresses the data by splitting the index key into two parts; the leading group of columns, called the prefix entry (which are potentially shared across multiple key values), and the suffix columns (which is unique to every index key). <br>
Key values in index entries only `direct traffic’; Insert/delete must be suitably modified.

## Hash versus Tree Index (Question 4.2)

What are three reasons to use a B+ Tree instead of a Hash Index?


_Answer:_


Insert/delete at $log_F N$ cost; keep tree heightbalanced.(F = fanout, N = # leaf pages)<br>
Minimum 50% occupancy (except for root). <br>
Supports equality and range-searches efficiently.

## Query Processing (Question 4.3)

Give three simple examples of rewriting a query to produce an equivalent query.

_Answer_

- R ⋈ S is the same as S ⋈ R
- $\sigma $(R ⋈ S) is the same at $\sigma $(R) ⋈ $\sigma $(S)
- $\delta  $(R ⋈ S) = $\delta$ (R) ⋈ $\delta $(S), where $\delta $ is the distinct operator.

## Buffer Management (Question 4.4)

Briefly explain _force, no force, steal_ and _no steal_ buffer management policies.

_Answer_

Force policy – make sure that every update is on disk before commit.<br>
No force - A page, whether modified or not, stays in the buffer as long as it is still needed. Only if it becomes the replacement victim it will be written to disk. <br>
Steal -  pages can be written out to disk even if the transaction having modified the pages is still active. <br>
No Steal policy – don’t allow buffer-pool frames with uncommited updates to overwrite committed data on disk.<br>

## Hash Index (Question 4.5)

<img src="./hash_index.jpeg">

The image below shows the state of a hash index.
- The table size is 11.
- The hashing algorithm is modulo table size.
- The rehash function is linear probing.

Show the new status of the cache after adding:
- 24
- 30

_Answer:_

24: 4<br>
30: 8

## Join Optimization (Question 4.6)

Briefly explain a Sort-Merge Join. When would you use a Sort-Merge instead of a Hash-Join and vice-versa.

_Answer:_

The sort-merge join combines two sorted lists like a zipper. Both sides of the join must be sorted by the join predicates. The key idea of the sort-merge algorithm is to first sort the relations by the join attribute, so that interleaved linear scans will encounter these sets at the same time.
SORT-MERGE joins can be used only for equijoins, it requires temporary segments for sorting. A sort-merge join may be performed when there is no index on the join columns, the selectivity of the join columns is low, or the clustering factor is high

## Deadlock (Question 4.7)

Briefly explain two techniques a transaction manager may use to break deadlocks?

_Answer:_

Use the waifor graph and explicitly abort a transaction.<br>
The more common approach is to use ”timeouts,” retry the transaction and hope the problem goes away.

## RAID (Question 4.8)

What is a major advantage of RAID-5 over RAID-0 and RAID-1?

_Answer:_

Read data transactions are very fast while write data transactions. If a drive fails, you still have access to all data, even while the failed drive is being replaced and the storage controller rebuilds the data on the new drive.

## Disk Addressing (Question 4.9)

Briefly explain CHS Addressing for disks.

_Answer:_


CHS addressing is the process of identifying individual sectors on a disk by their position in a track, where the track is determined by the head and cylinder numbers. <br>
CHS addresses can be converted to LBA addresses using the following formula:
LBA = (( C x HPC ) + H ) x SPT + S – 1
- C, H and S are the cylinder number, the head number, and the sector number<br>
- LBA is the logical block address<br>
- HPC is the number of heads per cylinder<br>
- SPT is the number of sectors per track

## ETag Processing (Question 4.10)

Briefly explain ETag processing. Is ETag processing more similar to pessimistic or optimistic transaction?

_Answer:_

Common methods of ETag generation include using a collision-resistant hash function of the resource's content, a hash of the last modification timestamp, or even just a revision number.<br>
- The server maintains an ETag (Entity Tag) for each resource.
- Every time a resource’s state changes, the server computes a new ETag.
- The server includes the ETag in the header when returning data to the client.
- The server may optionally maintain and return “Last Modified” time for resources.
- If-Match – value of a previous calls ETag response used with a PUT or DELETE. Server should only act if nobody else has modified the resource since you last fetched it. Otherwise provides a 412.
- If-Modified-Since – value of a previous Last-Modified response used with a GET. Server should only provide a response if the resource was modified since the timestamp submitted. Use in conjunction with If-None-Match in case the change occurred within the same second. Otherwise provide a 304.
- If-None-Match – value of a previous calls ETag response used with a GET. Server should only provide a response if the ETag doesn’t match, i.e. the resource has been altered. Otherwise provide a 304.
- If-Unmodified-Since – value of a previous Last-Modified response used with a PUT or DELETE. Server should only act if nobody else has modified the resource since you last fetched it. Otherwise provides a 412

It is more similar to optimistic transcations

# Bonus Question

Who is the greatest hockey player of all time?

_Answer:_

Bobby Orr